server/groundstation.py

```diff
@@ -82,10 +85,36 @@ def uav_thread(self):
             time.sleep(0.1)
 
     def image_thread(self):
-        while True:
-            run = self.image.retreive_images()
-            # print("Image:", run)
-            time.sleep(0.1)
+        if self.config["socket"]["connection"]:  # Initialize a socket connection
+            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
```

Comment:
Have you considered using UDP here? I think it's a more obvious use case for UDP here, but the part about UDP that worries me is that UDP doesn't retransmit lost packets, which is bad since we don't want to lose any image data that gets sent back. We may be able to use a combination of UDP and something else (perhaps the system could check for photos on the plane which haven't been processed at the end of the mission, and subsequently process them). 

Though I think TCP would be the better bet because of lost packet retransmition.

Reply:
That is exactly what I thought - UDP would be much simpler and effective in this scenario. 

You can simply send a UDP message after an image has been taken, and not worry about keeping a connection. For example, code to take images has been added, the function could just add a simple line and be done. However, with TCP, we probably need to do something more complicated. My idea with TCP was to keep a running list of images to be sent, and every time the socket connection "refreshes" (I'm thinking ~0.5 secs), it scans the list and sends any images which still need to be sent. This is probably the only way to send images reliably using TCP.

However, like you mentioned, UDP will not be reliable enough for our use-case. After sending, it is **impossible** to determine whether the send was successful or not. This is because of its nature as a whole - there is no 2-way handshake or anything. This means any packets "lost" are forever lost, and this is extremely bad considering we're not just connecting between two computers, but we have the Bullet's system as a source of error. 

This is why I chose TCP. If you want, we should talk about this more and make a solid decision - however, I feel like TCP is probably the best approach.